CONTRIBUTING: Package bumps should reference release notes

mweinelt opened the pull request:
Release notes for package upgrades are immensely helpful in judging whether the changes made to a package during a version bump are sensible, sufficient or complete.

This change clarifies that a good commit message should contain such a reference.

cc https://github.com/Mic92/nix-update/issues/87

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 commented on CONTRIBUTING.md:
```diff
@@ -51,7 +51,7 @@ See the nixpkgs manual for more details on [standard meta-attributes](https://ni
 In addition to writing properly formatted commit messages, it's important to include relevant information so other developers can later understand *why* a change was made. While this information usually can be found by digging code, mailing list/Discourse archives, pull request discussions or upstream changes, it may require a lot of work.
 
-For package version upgrades and such a one-line commit message is usually sufficient.
+For package version upgrades a simple commit message indicating the old and new version with a reference to the release notes or changelog are sufficient.
```
It is not quite clear that the headline is a must and the changelog is a should.

mweinelt replied:
I should probably change that to must be given, if one exists. Thanks for the pointer.

AndersonTorres replied:
The `meta.changelog` could be referenced for such things.

mweinelt replied:
Yeah, sure. Now only tools like `nix-update` and `nixpkgs-update` need to catch up in that regard.

SuperSandro2000 commented:
Can't we do a bot that fixes those up if a changelog exists?

Also people are going to chase me out of town with torches if I request this from them while reviewing.

mweinelt commented:
Well, here's two things in response to that:

- Having someone review the release notes on package bumps is important and providing a reference increases the likelihood of that happening.
- Putting the burden of retrieving the release notes for arbitrary package bump on the reviewer is something I strongly disagree with.

SuperSandro2000 commented:
> Having someone review the release notes on package bumps is important and providing a reference increases the likelihood of that happening.

If we would make the changelog URL static without the version in it, then it would be super easy to get to the changelog regardless if people manually put it into the commit body or used tools that do that.

AndersonTorres reviewed (commented):
Some suggestions:

- For simple version updates/upgrades, the ones that merely changes version and checksum, a one-line commit and a link to the changelog (when possible - some projects do not have changelog files) should suffice.
- Things that change the whole logic of Nix expression - such like "the build system changed from autotools to meson" - should be included in the commit messages.

mweinelt commented:
> > Having someone review the release notes on package bumps is important and providing a reference increases the likelihood of that happening.
> 
> If we would make the changelog URL static without the version in it, then it would be super easy to get to the changelog regardless if people manually put it into the commit body or used tools that do that.

Ideally `meta.changelog` gains more popularity, so we can offload the commit messages mostly to tools like `nix-update` and `nixpkgs-update`.

mweinelt committed:
CONTRIBUTING: Reference release notes in package bumps

This change improves the recommendation for good commit messages to
include release notes on package bumps.

Including the release notes increases the likelihood that they will be
taken into consideration during review. Having them included in the
review is important to be able to judge whether changes made during a
version bump are sensible, sufficient or complete.

The burden of retrieving the release notes for arbitrary package bumps
should not rest on the relatively small group of reviewers.

Signed-off-by: Martin Weinelt <hexa@darmstadt.ccc.de>

mweinelt commented:
Updated the wording taking some feedback into consideration. PTAL.
